numpy/_core/src/common/npy_hashtable.c

```diff
@@ -14,8 +14,10 @@
 
 #include "templ_common.h"
 #include "npy_hashtable.h"
+#include "common.h"
+#include "npy_pycompat.h"
 
-
+#ifndef Py_GIL_DISABLED
```

Comment:
I think we should have a comment here about the difference, and about how ideally we have a solution that works for both builds yet does not reduce performance when the GIL is enabled.

Note that ideally we indeed do not have our own hash tables - @seberg and I discussed other options at length when this code was added, but couldn't find an alternative that gave the same improvement.

Reply:
Right, about the comment: Maybe it makes sense to add something like:
```
/* Once cheap locking is exposed in Python, this should be used instead see gh-... */
```

>> Ufuncs never get deallocated

> I do not think we should assume th

Yeah, we shouldn't. But, I think it was more of an observation here rather than very important for the change.  I think the change actually prepares for making things more resiliant to an `ArrayMethod` getting deleted (ufunc lifetime shouldn't matter much, as it cannot be deallocateed while in execution).

Although, I guess the observation is still true in NumPy, so deallocating ufuncs is possible not tested (well), unfortunately.